File: services/geospatial-intelligence-service/src/geospatial_base/handlers/features.rs
use axum::{Json, extract::State};

use crate::{
    AppState,
    domain::engine::{clustering, routing, spatial_query},
    handlers::{ServiceResult, bad_request, db_error, load_all_layers, not_found},
    models::spatial_index::{
        ClusterRequest, ClusterResponse, RouteRequest, RouteResponse, SpatialOperation,
        SpatialQueryRequest, SpatialQueryResponse,
    },
};

pub async fn query_features(
    State(state): State<AppState>,
    Json(request): Json<SpatialQueryRequest>,
) -> ServiceResult<SpatialQueryResponse> {
    if matches!(
        request.operation,
        SpatialOperation::Within | SpatialOperation::Intersects
    ) && request.bounds.is_none()
    {
        return Err(bad_request(
            "bounds are required for within/intersects queries",
        ));
    }
    if matches!(
        request.operation,
        SpatialOperation::Nearest | SpatialOperation::Buffer
    ) && request.point.is_none()
    {
        return Err(bad_request("point is required for nearest/buffer queries"));
    }

    let layers = load_all_layers(&state.db)
        .await
        .map_err(|cause| db_error(&cause))?;
    let layer = layers
        .into_iter()
        .find(|layer| layer.id == request.layer_id)
        .ok_or_else(|| not_found("layer not found"))?;
    Ok(Json(spatial_query::execute(&layer, &request)))
}

pub async fn cluster_features(
    State(state): State<AppState>,
    Json(request): Json<ClusterRequest>,
) -> ServiceResult<ClusterResponse> {
    let layers = load_all_layers(&state.db)
        .await
        .map_err(|cause| db_error(&cause))?;
    let layer = layers
        .into_iter()
        .find(|layer| layer.id == request.layer_id)
        .ok_or_else(|| not_found("layer not found"))?;
    Ok(Json(clustering::cluster(&layer, &request)))
}

pub async fn route_features(
    State(_state): State<AppState>,
    Json(request): Json<RouteRequest>,
) -> ServiceResult<RouteResponse> {
    if request.origin == request.destination {
        return Err(bad_request("origin and destination must differ"));
    }
    Ok(Json(routing::route(&request)))
}

File: services/ontology-exploratory-analysis-service/internal/handlers/geospatial/features.go
// Spatial-feature handlers ported 1:1 from
// services/ontology-exploratory-analysis-service/src/geospatial/geospatial_base/handlers/features.rs
// (S8 / ADR-0030 — geospatial-intelligence-service absorbed). The
// route_features handler is intentionally not ported in this slice —
// see ROADMAP OEA-5+ where routing.rs lands its own engine and HTTP
// surface. Only query_features and cluster_features (the predicate +
// clustering surface for OEA-4) are wired here.
package geospatial

import (
	"encoding/json"
	"net/http"

	"github.com/google/uuid"

	"github.com/openfoundry/openfoundry-go/services/ontology-exploratory-analysis-service/internal/domain/spatial"
	"github.com/openfoundry/openfoundry-go/services/ontology-exploratory-analysis-service/internal/models"
)

// QueryFeatures mirrors `handlers::features::query_features`.
func (s *AppState) QueryFeatures(w http.ResponseWriter, r *http.Request) {
	var req models.SpatialQueryRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	switch req.Operation {
	case models.SpatialOperationWithin, models.SpatialOperationIntersects:
		if req.Bounds == nil {
			writeError(w, http.StatusBadRequest, "bounds are required for within/intersects queries")
			return
		}
	case models.SpatialOperationNearest, models.SpatialOperationBuffer:
		if req.Point == nil {
			writeError(w, http.StatusBadRequest, "point is required for nearest/buffer queries")
			return
		}
	}

	layer, ok, err := s.findLayer(r, req.LayerID)
	if err != nil {
		s.dbError(w, err)
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "layer not found")
		return
	}

	writeJSON(w, http.StatusOK, spatial.Execute(layer, req))
}

// ClusterFeatures mirrors `handlers::features::cluster_features`.
func (s *AppState) ClusterFeatures(w http.ResponseWriter, r *http.Request) {
	var req models.ClusterRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	layer, ok, err := s.findLayer(r, req.LayerID)
	if err != nil {
		s.dbError(w, err)
		return
	}
	if !ok {
		writeError(w, http.StatusNotFound, "layer not found")
		return
	}

	writeJSON(w, http.StatusOK, spatial.Cluster(layer, req))
}

// findLayer mirrors the Rust helper inlined in features.rs:
// load all layers, then look up by id in-memory. The Rust impl uses
// `load_all_layers` rather than `load_layer_row` to keep the lazy
// indexer hot — preserve that semantics here verbatim.
func (s *AppState) findLayer(r *http.Request, id uuid.UUID) (models.LayerDefinition, bool, error) {
	layers, err := LoadAllLayers(r.Context(), s.DB)
	if err != nil {
		return models.LayerDefinition{}, false, err
	}
	for _, layer := range layers {
		if layer.ID == id {
			return layer, true, nil
		}
	}
	return models.LayerDefinition{}, false, nil
}
